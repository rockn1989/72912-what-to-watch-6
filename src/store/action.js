export const ActionType = {
  SET_ACTIVE_GENRE: `films/set_genre`,
  SHOW_MORE: `films/show_more`,
  SET_FILMS_COUNTER: `films/set_films_counter`,
  LOAD_FILMS: `films/load_films`,
  LOAD_FILM: `films/id/load_film`,
  AUTHORIZED: `user/authorized`,
  SET_USER_INFO: `user/set_user_info`,
  SEND_COMMENT: `user/send_comment`,
  REDIRECT_TO_ROUTE: `redirect_to_route`,
  SEND_FORM_DATA: `form/send_form_data`,
  SHOW_ERROR: `show_error`
};

export const setGenreAction = (genre) => ({
  type: ActionType.SET_ACTIVE_GENRE,
  payload: genre
});

export const loadFilmsAction = (films) => ({
  type: ActionType.LOAD_FILMS,
  payload: films
});

export const loadFilmAction = (film) => ({
  type: ActionType.LOAD_FILM,
  payload: film
});

export const showMoreAction = () => ({
  type: ActionType.SHOW_MORE
});

export const setFilmsCounterAction = (counters) => ({
  type: ActionType.SET_FILMS_COUNTER,
  payload: counters
});

export const authorizationAction = (status) => ({
  type: ActionType.AUTHORIZED,
  payload: status
});

export const setUserInfoAction = (data) => ({
  type: ActionType.SET_USER_INFO,
  payload: data
});

export const redirectAction = (url) => ({
  type: ActionType.REDIRECT_TO_ROUTE,
  payload: url
});

export const sendUserCommentAction = (data) => ({
  type: ActionType.SEND_COMMENT,
  payload: data
});

export const sendFormDataAction = (bool) => ({
  type: ActionType.SEND_FORM_DATA,
  payload: bool
});

export const showErrorAction = (bool) => ({
  type: ActionType.SHOW_ERROR,
  payload: bool
});

