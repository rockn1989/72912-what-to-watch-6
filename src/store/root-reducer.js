import {combineReducers} from 'redux';

import {filmsData} from './films-data/films-data';
import {filmData} from './film-data/film-data';
import {user} from './user/user';
import {formStatus} from './form-status/form-status';
import {errorStatus} from './error-status/error-status';

export const NameSpace = {
  FILMS_DATA: `FILMS_DATA`,
  FILM_DATA: `FILM_DATA`,
  USER: `USER`,
  FORM_STATUS: `FORM_STATUS`,
  ERROR_STATUS: `ERROR_STATUS`
};

export default combineReducers({
  [NameSpace.FILMS_DATA]: filmsData,
  [NameSpace.FILM_DATA]: filmData,
  [NameSpace.USER]: user,
  [NameSpace.FORM_STATUS]: formStatus,
  [NameSpace.ERROR_STATUS]: errorStatus,
});

// export default combineReducers({
//   filmsData,
//   genre,
//   user,
//   formStatus,
//   errorStatus,
// });
