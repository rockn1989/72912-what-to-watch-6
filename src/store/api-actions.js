import {ActionCreator} from "./action";

export const loadFilmsList = () => (dispatch, _getState, api) => {
  api.get(`/films`)
    .then(({data}) => {
      dispatch(ActionCreator.loadFilms(data));
      dispatch(ActionCreator.filterFilms());
    });
};

export const loadFilm = (id) => (dispatch, _getState, api) => {
  api.get(`/films/${id}`)
    .then(({data}) => {
      dispatch(ActionCreator.loadFilm(data));
    })
    .catch(() => {
      dispatch(ActionCreator.redirect(`/page-not-found`));
    });
};

export const login = () => (dispatch, _getState, api) => {
  api.get(`/login`)
    .then(({data, status}) => {
      if (status !== 401) {
        dispatch(ActionCreator.authorization(true));
        dispatch(ActionCreator.setUserInfo(data));
      }
    });
};

export const logout = () => (dispatch, _getState, api) => {
  api.get(`/logout`)
    .then(({status}) => {
      if (status === 200) {
        dispatch(ActionCreator.authorization(false));
      }
    });
};

export const sendLogin = ({email, password}) => (dispatch, _getState, api) => {
  api.post(`/login`, {
    email,
    password
  }).then(({data, status}) => {
    if (status === 200) {
      dispatch(ActionCreator.authorization(true));
      dispatch(ActionCreator.setUserInfo(data));
      dispatch(ActionCreator.redirect(`/`));
    } else {
      dispatch(ActionCreator.authorization(false));
    }
  });
};

export const sendComment = ({id, comment, rating}) => (dispatch, _getState, api) => {
  dispatch(ActionCreator.sendFormData(false));
  api.post(`/comments/${id}`, {
    rating,
    comment
  }).then(() => {
    dispatch(ActionCreator.sendFormData(true));
    dispatch(ActionCreator.redirect(`/films/${id}`));
  })
  .catch(() => {
    dispatch(ActionCreator.showError(true));
  });
};
