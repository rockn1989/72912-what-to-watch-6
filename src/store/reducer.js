import {ActionType} from './action';
import {AUTH, NO_AUTH} from '../const';

const initialState = {
  genre: `All genres`,
  film: {},
  films: [],
  filtredFilmsList: [],
  filmsCounter: [],
  authorizationStatus: NO_AUTH,
  userInfo: {
    avatarUrl: `img/avatar.jpg`
  }
};

const reducer = (state = initialState, action) => {

  switch (action.type) {
    case ActionType.LOAD_FILMS:
      return {
        ...state,
        films: action.payload
      };

    case ActionType.LOAD_FILM:
      return {
        ...state,
        film: action.payload
      };

    case ActionType.SET_ACTIVE_GENRE:
      return {
        ...state,
        genre: action.payload
      };

    case ActionType.FILTER_FILMS:
      return {
        ...state,
        filtredFilmsList: state.genre === initialState.genre ? [...state.films] : state.films.filter((film) => film.genre === state.genre)
      };

    case ActionType.SHOW_MORE:
      const genreCounter = state.filmsCounter.filter(({name}) => name === state.genre);
      const idx = state.filmsCounter.indexOf(...genreCounter);

      const newEl = {
        name: state.genre,
        counter: genreCounter[0].counter + 1
      };

      const newFilmsCounter = [...state.filmsCounter.slice(0, idx), newEl, ...state.filmsCounter.slice(idx + 1)];
      return {
        ...state,
        filmsCounter: [...newFilmsCounter]
      };

    case ActionType.SET_FILMS_COUNTER:
      const counters = [];
      for (let name of action.payload) {
        counters.push({
          name,
          counter: 0
        });
      }

      return {
        ...state,
        filmsCounter: [...counters]
      };

    case ActionType.RESET_FILMS_COUNTER:
      const resetFilmsCounter = state.filmsCounter.map((counter) => {
        return {
          name: counter.name,
          counter: 0
        };
      });
      return {
        ...state,
        filmsCounter: [...resetFilmsCounter]
      };

    case ActionType.AUTHORIZED:
      return {
        ...state,
        authorizationStatus: action.payload ? AUTH : NO_AUTH
      };

    case ActionType.SET_USER_INFO:
      const {avatar_url: avatar, ...restData} = {...action.payload};
      return {
        ...state,
        userInfo: {...restData, avatarUrl: avatar}
      };
  }
  return state;
};

export {reducer};
