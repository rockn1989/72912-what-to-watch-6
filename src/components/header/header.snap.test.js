import React from 'react';
import {render} from '@testing-library/react';
import {Router} from 'react-router-dom';
import {createMemoryHistory} from 'history';
import configureStore from 'redux-mock-store';
import Header from './header';
import {NameSpace} from '../../store/root-reducer';
import * as redux from 'react-redux';
const mockStore = configureStore({});

it(`Should Header render correctly`, () => {
  jest.spyOn(redux, `useSelector`);

  const store = mockStore({
    [NameSpace.USER]: {
      authorizationStatus: true,
      userInfo: {
        avatarUrl: `img/avatar.jpg`
      }
    }
  });

  const history = createMemoryHistory();

  const {container} = render(
      <redux.Provider store={store}>
        <Router history={history}>
          <Header addClassName={`movie-card__head`}/>
        </Router>
      </redux.Provider>
  );

  expect(container).toMatchSnapshot();
});
