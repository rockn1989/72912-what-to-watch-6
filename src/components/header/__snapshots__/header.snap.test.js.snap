// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Header render correctly 1`] = `
<div>
  <header
    class="page-header undefined"
  >
    <div
      class="logo"
    >
      <a
        class="logo__link"
        href="/"
      >
        <span
          class="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          class="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          class="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <div
      class="user-block"
    >
      <div
        class="user-block__avatar"
      >
        <img
          alt="User avatar"
          height="63"
          src="img/avatar.jpg"
          width="63"
        />
      </div>
    </div>
  </header>
</div>
`;
